# IgSeq Utilities

**Work in Progress**

A command-line tool for various common Ig-Seq tasks.  These are heavily biased
toward the peculiarities of our protocol and for rhesus macaque antibody
sequences.  Your mileage may vary.

The `igseq` command is organized into subcommands, grouped into two
categories: early read processing tasks (demultiplex, trim, merge, etc.), and
various convenience tools (IgBLAST this against that, what database has the
closest V gene, etc.).

Read processing subcommands:

 * getreads: Run [bcl2fastq] with an autogenerated sample sheet and some
   customized settings to write Undetermined I1/R1/R2 fastq.gz files.
 * demux: Demultiplex the I1/R1/R2 files according to per-sample barcodes
 * phix: Map reads left unassigned post-demux to the PhiX genome for
   troubleshooting
 * trim: Run [cutadapt] to remove adapter/primer/barcode and low-quality
   sequences on a per-sample basis
 * merge: Merge R1/R2 for each sample with [PEAR].

Each step in the read processing produces a read counts summary CSV table
`<step>.counts.csv` and has default output paths derived from the inputs, and
most can work per-sample or per-directory, so it's easy to chain together the
steps for a given run:

    igseq getreads /seq/runs/211105_M05588_0469_000000000-JWV49
    igseq demux -s samples.csv analysis/reads/211105_M05588_0469_000000000-JWV49
    igseq phix analysis/demux/211105_M05588_0469_000000000-JWV49
    igseq trim -s samples.csv analysis/demux/211105_M05588_0469_000000000-JWV49
    igseq merge analysis/trim/211105_M05588_0469_000000000-JWV49

[bcl2fastq]: https://support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html
[cutadapt]: https://cutadapt.readthedocs.io
[PEAR]: https://cme.h-its.org/exelixis/web/software/pear/
